bot/exts/moderation/infraction/management.py

```diff
@@ -201,29 +201,34 @@ async def infraction_edit(
     # region: Search infractions
 
     @infraction_group.group(name="search", aliases=('s',), invoke_without_command=True)
-    async def infraction_search_group(self, ctx: Context, query: t.Union[UserMention, Snowflake, str]) -> None:
+    async def infraction_search_group(self, ctx: Context, query: t.Union[MemberOrUser, Snowflake, str]) -> None:
```

Comment:
It just occurred to me why this didn't originally use the normal converter. The problem is that discord.py's converters also support username and nickname strings (without a discriminator). This makes it impossible to search a reason for a string which happens to be a user's name. In practice, users are rarely, if ever specified this way, so it's preferable to prioritise the reason search.

Reply:
That makes sense.

Would it be more logical to instead use `UserMentionOrId` here, since then that will remove the need of the `Snowflake` converter? We'd likely have to add a try/except around the call to the UserConverter though -- so `UserMentionOrId` becoming something like
```py
    async def convert(self, ctx: Context, argument: str) -> t.Union[discord.User, int]:
        """Convert the `arg` to a `discord.User` or `int` in the case of a deleted account."""
        match = self._get_id_match(argument) or RE_USER_MENTION.match(argument)

        if match is not None:
            try:
                return await super().convert(ctx, argument)
            except commands.UserNotFound:
                return int(argument)
        else:
            raise BadArgument(f"`{argument}` is not a User mention or a User ID.")
```The converter currently isn't used anywhere so this change wouldn't break anything elsewhere.

The only downside to this I can see is an id being passed which is a message id or something and so the passed value wouldn't work for user operations -- although this is already the case for `Snowflake` and there aren't any user operations done in `infraction_search_group` anyway so should be fine). Would just mean the bot says "No infractions could be found for that query." (as it does currently if you pass a message id since AFAIK there's no real way to differentiate between id types).

Reply:
I also just found this so will fix that as well I guess (typehint says it'll be a list but it's actually a tuple)
![image](https://user-images.githubusercontent.com/47674925/129475450-fd1745af-ba63-49ba-8657-d226c4d53bcc.png)

Reply:
> Would it be more logical to instead use UserMentionOrId here, since that will remove the need of the Snowflake converter? We'd likely have to add a try/except around the call to the UserConverter though to still allow for searching of deleted accounts etc. -- so UserMentionOrId would become something like:

The `Snowflake` converter has some code that improves the likelihood of a converted input actually being a snowflake. This would further disambiguate snowflakes from reasons that look like snowflakes. However, I'm not sure how often such a long number is included in a reason; it's probably unnecessary to have this extra validation for this command.

That being said, keeping the converters separate would make `UserMentionOrId` more versatile for any future use cases. If it were to fall back to returning the snowflake, then it should return a `discord.Object` instead of an integer.

> Would just mean the bot says "No infractions could be found for that query.", as it does currently if you pass a message id -- no real way to fix this since AFAIK you can't differentiate between p.e a message id and a user id.

Yes, that's just something we have to accept.

Reply:
So are you suggesting using `UserMentionOrId` (not `Snowflake`), but changing the convert method to return a `discord.Object` when the call to UserConverter fails?

Reply:
Either that or keeping them separate like `Union[UserMentionOrId, Snowflake, str]` so that `UserMentionOrId` remains versatile. I think the latter would be better. What do you think?

Reply:
Yeah, I think keeping it versatile is better actually. I'll do that :+1:

Reply:
@MarkKoz A bit of a delay because I was away camping but have just [pushed a commit](https://github.com/python-discord/bot/pull/1742/commits/3ce1272c7e556e4b1cb8f330fd2553f20842c061) for this :+1:

Reply:
My comment was only on `infraction_search_group`. `search_user` did not need to be updated because it only accepts users; there's no risk of ambiguity between names and reasons.

Reply:
If I don't change `search_user` then I can a lint error on L211 since I'm passing a `UserMentionOrId` to a `MemberOrUser` parameter. In reality `UserMentionOrId` gives a `discord.User` so it isn't actually an issue but the linter doesn't pick up on that.

Am I okay to just ignore this linting error and revert `search_user` back to the `MemberOrUser` converter?

Reply:
Yes just ignore it.

Reply:
A type checker like mypy isn't used, so it will not fail the CI step. There is a check that a type annotation exists, but it doesn't not enforce type correctness.

Reply:
It has been [committed](https://github.com/python-discord/bot/pull/1742/commits/c88698727cf56b1311e7219ec2dfa06c3a7863ce)